android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireActivity.kt

```diff
@@ -93,117 +100,117 @@ open class QuestionnaireActivity : BaseMultiLanguageActivity(), View.OnClickList
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
+    requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
     setContentView(R.layout.activity_questionnaire)
-    if (!intent.hasExtra(QUESTIONNAIRE_ARG_FORM)) {
-      showToast(getString(R.string.error_loading_form))
-      finish()
+
+    val formName = intent.getStringExtra(QUESTIONNAIRE_ARG_FORM)!!
+    clientIdentifier = intent.getStringExtra(QUESTIONNAIRE_ARG_PATIENT_KEY)
+    intent.getStringExtra(QUESTIONNAIRE_ARG_TYPE)?.let {
+      questionnaireType = QuestionnaireType.valueOf(it)
     }
 
     val loadProgress = showProgressAlert(this, R.string.loading)
 
-    lifecycleScope.launch { questionnaireViewModel.libraryEvaluator.initialize() }
+    lifecycleScope.launch {
+      loadQuestionnaireAndConfig(formName)
 
-    clientIdentifier = intent.getStringExtra(QUESTIONNAIRE_ARG_PATIENT_KEY)
-    immunizationId = intent.getStringExtra(ADVERSE_EVENT_IMMUNIZATION_ITEM_KEY)
-
-    lifecycleScope.launchWhenCreated {
-      readOnly = intent.getBooleanExtra(QUESTIONNAIRE_READ_ONLY, false)
-      editMode = intent.getBooleanExtra(QUESTIONNAIRE_EDIT_MODE, false)
+      async(Dispatchers.Default) { questionnaireViewModel.libraryEvaluator.initialize() }
```

Comment:
Can we inject and use `DefaultDispatcherProvider` here so this can be replaced with `TestDispatcher` in tests.

Reply:
`async` can be replaced with `withContext`

Reply:
Is the task CPU intensive? Hence the reason for using `Default` Dispatcher?

Reply:
Would use the withContext and DefaultDispatcherProvider. Yeah. This starts loading the FHIR and CQL classes and resources which keeps the Main thread engaged unnecessarily.

Reply:
@ellykits [Doumentation](https://proandroiddev.com/async-operations-with-kotlin-coroutines-part-1-c51cc581ad33) says "withContext function can be used to switch the context of a coroutine and allow the execution of a block within the new context. This function suspends until the block completes and returns the result." But I actually want to unblock operation and don`t care about the outcome of the block

Reply:
Yes that is correct, as long as you provide the correct dispatcher there should be no reason to worry. Using `async` has the same effect in this context.

Reply:
If there is no need for a return value then use `launch` instead

Reply:
From the code snippet, there are 3 coroutine tasks launched in the same scope.

Reply:
You can split the tasks launch the one you want to run in parallel with the others separately. The point is, it is not recommended to use `async` when the result deferred is not used.

Reply:
My question on using the `Dispatchers.Default` as the dispatcher, from your response

>Yeah. This starts loading the FHIR and CQL classes and resources which keeps the Main thread engaged unnecessarily.

sounds like the operation is mostly I/O and is not computation intensive in that case use `dispatcher.io()` thread which is better optimized for such use cases.

Reply:
done